hw14.1_todo_list/src/index.ts

```diff
@@ -0,0 +1,194 @@
+// Окремо необхідно розширити поведінку списку та додати можливість пошуку нотатка за ім'ям або змістом.
+// Також окремо необхідно розширити список можливістю сортування нотаток за статусом або часом створення.
+
+enum StatusTodoListEnum {
+  New = 'NEW',
+  IsPerformed = 'IS PERFORMED',
+  Executed = 'EXECUTED'
+}
+
+enum TypeOfTaskEnum {
+  Default = 'DEFAULT',
+  EditConfirm = 'EDIT CONFIRMATION'
+}
+
+interface ITodoTask {
+  id: number;
+  name: string;
+  description: string;
+  dateCreated: Date;
+  dateEdited?: Date | null;
+  editing: boolean;
+  status: StatusTodoListEnum;
+  typeOfTask: TypeOfTaskEnum;
+}
+
+interface ITodoList {
+  addNewTask(task: ITodoTask): void;
+  removeTask(taskId: number): void;
+  editingTask(taskId: number, updatedTask: Partial<ITodoTask>): void;
+  getInformationTask(taskId: number): ITodoTask | undefined;
+  getAllOfTasks(): ITodoTask[];
+  getTotalTasksCount(): number;
+  markTaskAsExecuted(taskId: number): void;
+  getExecutedTaskCount(): number;
+  getTotalTasksCount(): number;
+  searchTasks(query: string): ITodoTask[];
+  sortTasks(by: 'status' | 'dateCreated'): ITodoTask[]
+}
+
+class TodoList implements ITodoList {
+  private tasks: ITodoTask[] = [];
+  private nextId: number = 1;
+
+  
+  //Add a new task
+  public addNewTask(task: ITodoTask): void {
+    task.id = this.nextId++;
+    task.dateCreated = new Date();
+    task.dateEdited = null;
+    this.tasks.push(task)
+  };
+  
+  //Remove a task
+  public removeTask(taskId: number): void {
+    this.tasks = this.tasks.filter(item => item.id !== taskId)
+  };
+
+  // Editing a task  
+  public editingTask(taskId: number, updatedTask: Partial<ITodoTask>): void {
+    const task = this.tasks.find(task => task.id === taskId);
+    if (task) {
+      if (task.typeOfTask === TypeOfTaskEnum.EditConfirm) {
+        task.editing = true;
+      }
+      Object.assign(task, updatedTask);
+      task.dateEdited = new Date();
+      task.editing = false;
+    }
+  };
+
+  // Get all of information about some task  
+  public getInformationTask(taskId: number): ITodoTask | undefined {
+    return this.tasks.find(task => task.id === taskId);
+  };
+  
+  // Get all of tasks
+  public getAllOfTasks(): ITodoTask[] {
+    return this.tasks;
+  };
+
+  //Mark a task as executed
+  public markTaskAsExecuted(taskId: number): void {
+    const task = this.tasks.find(task => task.id === taskId);
+    if(task) {
+      task.status = StatusTodoListEnum.Executed;
+    }
+  }
+
+  //Count all of executed tasks
+  public getExecutedTaskCount(): number{
+    return this.tasks.filter(task => task.status !== StatusTodoListEnum.Executed).length;
+  }
+
+  //Count all of tasks
+  public getTotalTasksCount(): number {
+    return this.tasks.length;
+  }
+
+  //Search tasks by name and description
+  public searchTasks(query: string): ITodoTask[] {
```

Comment:
This method needs to be added to a separate entity. Please ensure that the requirements are carefully followed.